signalclient.go

```diff
@@ -348,18 +350,17 @@ func (c *SignalClient) handleResponse(res *livekit.SignalResponse) {
 func (c *SignalClient) readWorker() {
 	defer func() {
 		c.isStarted.Store(false)
-		if c.OnClose != nil {
-			c.OnClose()
-		}
+		c.conn.Store((*websocket.Conn)(nil))
+		close(c.readerClosedCh)
 	}()
 	if pending := c.pendingResponse; pending != nil {
 		c.handleResponse(pending)
 		c.pendingResponse = nil
 	}
-	for !c.isClosed.Load() {
+	for {
 		res, err := c.readResponse()
 		if err != nil {
-			if !isIgnoredWebsocketError(err) && !c.isClosed.Load() {
```

Comment:
we'll also need to make sure readResponse returns EOF if it's expected close. otherwise it'll keep spamming logs about error while reading.

Reply:
Will look at this.

Reply:
On second thought, this  does not do anything different. A unilateral close (by calling Close() method) will close the underlying conn and read will fail with `reading a closed conn`. And this loop will exit/return.

Before this change, the code was only checking `isClosed` which is set by the unilateral `Close()`.

The remote close will be an error returned by `ReadMessage`. Yes, this change will log one message more because it will log an error on an unilateral `Close()` too. But, nothing more.

Once an error is returned, the loop is exited.

Am I missing something?

Reply:
maybe I'm reading this incorrectly.. [here](https://github.com/livekit/server-sdk-go/pull/351/files#diff-db22f9da506d0e6a0874eee29d06d041ca0afa8c9d74681a846af283252f648fL266) it would previously return io.EOF when `Close` has been called due to the `c.isClosed.Load()` check.. currently I think it'll go into `errors.New("cannot read response before join")` ?

It seems that we should not return an error if Close has been called.

Reply:
That's correct @davidzhao . You are reading it right. But, what I was saying is that `Close()` in this PR sets the `conn` to `nil` (the other discussion about `atomic.Value == nil`, I will change that to an `atomic.Pointer`). So, yes `readResponse` will return `errors.New("cannot read response before join")`. 

But, the `readLoop` will exit because of that error return.

Only difference is that previously, the read loop would not have logged an error before exiting as it was checking for `isClosed` and with this PR, it will log an error before exiting. So, what I was mentioning was that I could not spot any runaway loops repeating logging an error.

Reply:
Sorry, I didn't mean it would be in a loop. Just that in server environments, it would be good to avoid logging errors when the user has closed the connection.

Reply:
Got it, I misunderstood. I will look at adding a change to filter out the error

Reply:
Addressed in this PR - https://github.com/livekit/server-sdk-go/pull/353